cmd/tether/attach.go

```diff
@@ -0,0 +1,369 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"syscall"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/vmware/vic/pkg/dio"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/net/context"
+)
+
+const (
+	attachChannelType = "attach"
+)
+
+var (
+	Signals = map[ssh.Signal]int{
+		ssh.SIGABRT: 6,
+		ssh.SIGALRM: 14,
+		ssh.SIGFPE:  8,
+		ssh.SIGHUP:  1,
+		ssh.SIGILL:  4,
+		ssh.SIGINT:  2,
+		ssh.SIGKILL: 9,
+		ssh.SIGPIPE: 13,
+		ssh.SIGQUIT: 3,
+		ssh.SIGSEGV: 11,
+		ssh.SIGTERM: 15,
+		ssh.SIGUSR1: 10,
+		ssh.SIGUSR2: 12,
+	}
+)
+
+// PtyRequestMsg the RFC4254 struct
+type ptyRequestMsg struct {
+	Term     string
+	Columns  uint32
+	Rows     uint32
+	Width    uint32
+	Height   uint32
+	Modelist string
+}
+
+// WindowChangeMsg the RFC4254 struct
+type WindowChangeMsg struct {
+	Columns  uint32
+	Rows     uint32
+	WidthPx  uint32
+	HeightPx uint32
+}
+
+type signalMsg struct {
+	Signal string
+}
+
+var server attachServer
+
+type attachServer interface {
+	start() error
+	stop()
+}
+
+// conn is held directly as it's how we stop the attach server
+type attachServerSSH struct {
+	conn   *net.Conn
+	config *ssh.ServerConfig
+
+	enabled bool
+}
+
+// start is not thread safe with stop
+func (t *attachServerSSH) start() error {
+	if t == nil {
+		return errors.New("attach server is not configured")
+	}
+
+	if t.enabled {
+		return nil
+	}
+
+	// don't assume that the key hasn't changed
+	pkey, err := ssh.ParsePrivateKey(config.Key)
+	if err != nil {
+		detail := fmt.Sprintf("failed to load key for attach: %s", err)
```

Comment:
I'm wanting to both log the formatted string as well as construct an error from it so I think the only other choice is:
```
err := fmt.Errorf("...%s", err)
log.Error(err.Error())
return err
```

as I've no idea how complex the Error() call can get I went with the current approach.